### PR TITLE
chore: new 'types' module

### PR DESCRIPTION
**Pull request by alekst23**
## Description
(replaces PR https://github.com/defenseunicorns/leapfrogai/pull/963)

We need a better way to organize our type definitions. Currently we have the following issues:
- One file with a lot of definitions: `src/leapfrogai_api/backend/types.py`. Very long file with many types for different purposes defined together. Difficult to analyze and find code.
- Some types mix data models and complex function implementations in one class
- Types are not defined in a consistent manner

Good format to follow is the OpenAI types module:
https://github.com/openai/openai-python/tree/main/src/openai/types

### BREAKING CHANGES
All changes to type definitions should be applied to all relevant files as part of the PR.

### CHANGES
- A new module for `leapfrogai_api` is created: `types`.
- Python Module format allows for complex type hierarchy and selective importing of objects. 
- Creates a standardized way for application to define and use types.
- Consolidates types defined inline and in `src/leapfrogai_api/backend/types` to the new `types` module
- Consolidates constants to `src/leapfrogai_api/types/constants.py`

## Related Issue

Precursor to: 
https://github.com/defenseunicorns/leapfrogai/issues/948

## Checklist before merging

- [ ] Tests, documentation, ADR added or updated as needed
- [ ] Followed the [Contributor Guide Steps](https://github.com/defenseunicorns/leapfrogai/blob/main/.github/CONTRIBUTING.md)

### CODE REVIEW

**Review by justinthelaw — CHANGES_REQUESTED** [review]
Two integration tests are currently failing in this refactor; 1 related to runs and 1 related to auth, and both of them are based on imports: https://github.com/defenseunicorns/leapfrogai/actions/runs/10782682964/job/29903199093#step:12:53

Everything else looks good, now that the comments from the team are addressed! This is a great refactor - love the organization.

**Comment by alekst23** [comment]
@justinthelaw Fixed and updated from main. Waiting for the tests to run again.

**Comment by justinthelaw** [comment]
> @justinthelaw Fixed and updated from main. Waiting for the tests to run again.

Looks like there is still 1 failing integration test, but the debug info might be in-cluster.

File: src/leapfrogai_api/routers/openai/runs.py [hunk 1]
```diff
@@ -84,7 +84,12 @@ async def create_thread_and_run(
         )
 
     try:
-        return await request.generate_response(new_run, new_thread, session)
+        return await Composer().generate_response(
+            request=request,
```
Comment (gphorvath) [inline src/leapfrogai_api/routers/openai/runs.py:88]: Typing conflict ThreadRunCreateParamsRequest RunCreateParamsRequest.

File: src/leapfrogai_api/backend/composer.py [hunk 2]
```diff
@@ -235,70 +112,60 @@ def sort_by_created_at(msg: Message):
...
+            file_search: BetaThreadToolResourcesFileSearch = cast(
+                BetaThreadToolResourcesFileSearch, tool_resources.file_search
+            )
+            vector_store_ids: list[str] = cast(list[str], file_search.vector_store_ids)
+
+            for vector_store_id in vector_store_ids:
+                rag_results_raw: SingleAPIResponse[
+                    SearchResponse
+                ] = await query_service.query_rag(
+                    query=query_message.content,
```
Comment (gphorvath) [inline src/leapfrogai_api/backend/composer.py:146]: Type conflict `list[TextContentBlockParam]` vs `str`.

File: .github/workflows/pytest.yaml [hunk 3]
```diff
@@ -90,8 +90,6 @@ jobs:
         with:
           path: |
             ~/.cache/pip
-            **/src/leapfrogai_api
```
Comment (justinthelaw) [inline .github/workflows/pytest.yaml:93]: Make sure to remove these lines from the original cache-ing action in the previous job too!

File: src/leapfrogai_api/backend/constants.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# This file defines application constants not set by environment variables or configuration files.
+
+THIRTY_DAYS_SECONDS = 60 * 60 * 24 * 30  # in seconds
+
```
Comment (justinthelaw) [inline src/leapfrogai_api/backend/constants.py:4]: Should we also put the default K in here just to centralize the similarity search variable?

**Review by justinthelaw — DISMISSED** [review]
Super minor nit that can be accomplished in a separate branch (so you don't have to run tests again). Looks great!

File: src/leapfrogai_api/backend/constants.py (hunk 4 above)
Comment (gphorvath) [reply src/leapfrogai_api/backend/constants.py]: I added this!

File: .github/workflows/pytest.yaml (hunk 3 above)
Comment (gphorvath) [reply .github/workflows/pytest.yaml:93]: done

**Review by jalling97 — APPROVED** [review]
Presuming the remaining tests pass, this looks good to me!

**Review by justinthelaw — APPROVED** [review]

**Review by CollectiveUnicorn — APPROVED** [review]

**Pull request merged**